juju/juju
# 2.4 - Containers on Manually Provisioned Machines, on Providers Supporting Container Networking

## Description of change

When adding a container to a manually provisioned machine on a provider that supports container networking, such as MAAS, provisioning fails because the machine is not recognised.

This patch adds an additional check when allocating container addresses. If the machine is manually provisioned, we just fall back to DHCP without asking the provider for an address.

*Note for reviewers*: The functional change is in the second commit (https://github.com/juju/juju/pull/9287/commits/8ba8c82307e7841b430d28d4fb7a65fa30c7076d). The first commit includes receiver renamings that add mostly noise, for which the submitter apologises.

## QA steps

There are no unit tests covering this change because the functionality in question is tested using the dummy provider, which does not support container networking. The QA steps confirm desired operation.

- Bootstrap to MAAS.
- Deploy a Xenial machine that uses DHCP. Note the address it receives.
- Add manually provisioned machine, for example `juju add-machine ssh:ubuntu@10.0.0.76`
- `juju add-machine lxd:0` and ensure that there is no error.
- `juju add-machine`
- `juju add-machine lxd:1` and check that normal container machine provisioning works`

## Documentation changes

None.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1796106
